packages/messaging/android.js

```diff
@@ -0,0 +1,111 @@
+/**
+ * @module Android Messaging
+ *
+ * @description A wrapper for messaging on Android. See example usage in android.transport.js
+ */
+import { MissingHandler } from './messaging.js'
+
+/**
+ * @typedef {import("./messaging").MessagingTransport} MessagingTransport
+ */
+
+/**
+ * On Android, handlers are added to the window object and are prefixed with `ddg`. The object looks like this:
+ *
+ * ```typescript
+ * {
+ *     onMessage: undefined,
+ *     postMessage: (message) => void,
+ *     addEventListener: (eventType: string, Function) => void,
+ *     removeEventListener: (eventType: string, Function) => void
+ * }
+ * ```
+ *
+ * You send messages to `postMessage` and listen with `addEventListener`. Once the event is received,
+ * we also remove the listener with `removeEventListener`.
+ *
+ * @link https://developer.android.com/reference/androidx/webkit/WebViewCompat#addWebMessageListener(android.webkit.WebView,java.lang.String,java.util.Set%3Cjava.lang.String%3E,androidx.webkit.WebViewCompat.WebMessageListener)
+ * @implements {MessagingTransport}
+ */
+export class AndroidMessagingTransport {
+    /** @type {AndroidMessagingConfig} */
+    config
+    globals = {capturedHandlers: {}}
+    /**
+     * @param {AndroidMessagingConfig} config
+     */
+    constructor (config) {
+        this.config = config
+    }
+
+    /**
+     * Given the method name, returns the related Android handler
+     * @param {string} methodName
+     * @returns {AndroidHandler}
+     * @private
+     */
+    _getHandler (methodName) {
+        const androidSpecificName = this._getHandlerName(methodName)
+        if (!(androidSpecificName in window)) {
+            throw new MissingHandler(`Missing android handler: '${methodName}'`, methodName)
+        }
+        return window[androidSpecificName]
```

Comment:
What's the story around these handlers, do we need to attempt to capture them? or is that out of scope for this work?

I know that previously the exec-time of Android's JS couldn't be guaranteed to be 'first' - can it now?

Reply:
I thought about that, and even attempted a quick solution, but then I backtracked. It's not strictly in scope, and we should think this through a bit more, especially if we think of that as improving security, because it may not change anything. We would likely still do it for a variety of reasons, but we need a triage etc. I will scope a project to follow up.

Reply:
Nice - yeah with a follow-up project in place then we can approve this and move forward 💪🏻